### PR TITLE
Inconsistent Update Behavior for SCIM "/Users/{userId}"

### LINKED ISSUE #2671: Inconsistent update behavior for SCIM /Users/{userId}
### What version of UAA are you running?

latest 

### How are you deploying the UAA?

not relevant, but docker, cf deployment

### What did you do?

Update the origin of a specific useruser via SCIM API /Users/{userId} , either via PATCH or PUT.

### What did you expect to see? What goal are you trying to achieve with the UAA?

Consistent behavior that both http methods either allow or disallow the update of the origin of a user via scim.

### What did you see instead?

- PUT allows update of the origin of a user
- PATCH does not allow the update of the origin of a user

```sh
# create a new user
curl -X POST '<uaadomain>/Users' -H "Authorization: bearer $TOKEN" -H "Content-type: application/json" -d '{"userName": "user-does-not-exist",  "origin": "does-not-exist", "emails": [{"value": "test@email.com", "primary": false}]}'

# try to update origin via patch, returns successfully but with old origin
curl -X PATCH '<uaadomain>/Users/<userid>' -H "Authorization: bearer $TOKEN" -H "Content-type: application/json" -H 'If-Match: *' -d '{"origin":"still-not-existing"}'

# update origin via put, returns successfully with new origin
curl -X PUT '<uaadomain>/Users/<userid>' -H "Authorization: bearer $TOKEN" -H "Content-type: application/json" -H 'If-Match: *' -d '{"userName": "user-does-not-exist",  "origin": "still-does-not-exist", "emails": [{"value": "test@email.com", "primary": false}], "name": {"familyName":"familyName", "givenName":"givenName"}}'

```

Why ?

PATCH does not take into consideration the origin while applying the provided delta.

https://github.com/cloudfoundry/uaa/blob/4dae99997cb27c3fea97a07ced3a2a2402f2e037/model/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUser.java#L711

PUT simply takes the provided user from the body and updates

https://github.com/cloudfoundry/uaa/blob/develop/server/src/main/java/org/cloudfoundry/identity/uaa/scim/endpoints/ScimUserEndpoints.java#L262

Current documentation mentions origin as update field for both methods.

Further, this suffers from the same issue as #2576 , i.e., i can PUT a non-existing origin.

### PR DESCRIPTION
**Pull request by adrianhoelzl-sap**
See Issue #2671

### CODE REVIEW

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/scim/endpoints/ScimUserEndpointsMockMvcTests.java [hunk 1]
```diff
@@ -1068,6 +1069,23 @@ void testUpdateUser_No_Username_Returns_400() throws Exception {
...
+
+        // change origin
+        final String initialOrigin = scimUser.getOrigin();
+        scimUser.setOrigin(UUID.randomUUID().toString());
+
+        scimUser.setName(new ScimUser.Name("Some New", "Name"));
+
+        updateUser(scimReadWriteToken, HttpStatus.BAD_REQUEST.value(), scimUser);
+
+        // should work again when origin was not changed
```
Comment (swalchemist) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/scim/endpoints/ScimUserEndpointsMockMvcTests.java]: This last part doesn't seem necessary for the test, and should be covered by a different test.

File: server/src/test/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioningTests.java [hunk 2]
```diff
@@ -1170,6 +1200,30 @@ private static void addUser(final JdbcTemplate jdbcTemplate,
         jdbcTemplate.execute(addUserSql);
     }
 
+    private static void addUser(final JdbcTemplate jdbcTemplate,
```
Comment (swalchemist) [inline server/src/test/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioningTests.java]: Could this signature be simplified with a parameter class, like ScimUser? It's a large number of parameters.

**Review by swalchemist — CHANGES_REQUESTED** [review]
Am I reading this right, that PATCH will refuse to modify the origin with no error, but PUT will give an error if a different origin is given? That doesn't meet the need to give them the same behavior.

Also, will the API documentation be updated to indicate that the origin cannot be changed?

File: server/src/test/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioningTests.java (hunk 2 above)
Comment (adrianhoelzl-sap) [reply server/src/test/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioningTests.java]: fixed with 0dec2a7

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/scim/endpoints/ScimUserEndpointsMockMvcTests.java (hunk 1 above)
Comment (adrianhoelzl-sap) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/scim/endpoints/ScimUserEndpointsMockMvcTests.java]: fixed with 0dec2a7

**Review by swalchemist — APPROVED** [review]

**Review by swalchemist — APPROVED** [review]

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/scim/endpoints/ScimUserEndpointsMockMvcTests.java
Comment (swalchemist) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/scim/endpoints/ScimUserEndpointsMockMvcTests.java:1]: Much easier to read now. Thanks!

**Review by klaus-sap — APPROVED** [review]

**Pull request merged**